File: Practice/AllDomains/Tutorials/CrackingTheCodingInterview/StacksBalancedBrackets.js
/*	author:@shivkrthakur  */
process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();    
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

/////////////// ignore above this line ////////////////////

function main() {
    var t = parseInt(readLine());
    for(var a0 = 0; a0 < t; a0++){
        var expression = readLine();
        if(areBalancedBrackets(expression)) console.log("YES");
        else console.log("NO");
    }
}

var areBalancedBrackets = function(input){
    var flag = true;
    var stack = [];
    var len = input.length;
    for(var i = 0; i < len; i++)
    {
        var char = input[i];
        if(char === '(' || char === '{' || char === '['){
            stack.push(char);
        }
        else {
            if(stack.length > 0){
                if(   (char === ')' && stack[stack.length - 1] !== '(') 
                   || (char === '}' && stack[stack.length - 1] !== '{') 
                   || (char === ']' && stack[stack.length - 1] !== '[')){
                    flag = false;
                }
                var pop = stack.pop();
            }
            else flag = false;
        }
    }
    if(flag && stack.length > 0) flag = !flag;
    return flag;
}

File: Practice/AllDomains/Tutorials/CrackingTheCodingInterview/StacksBalancedBrackets.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
class Solution {

static void Main(String[] args) {
    int t = Convert.ToInt32(Console.ReadLine());
    Stack<char> stack1;

    for(int a0 = 0; a0 < t; a0++){
        string expression = Console.ReadLine();
        int expLen = expression.Length;
        if(expLen % 2 != 0)
        {
            Console.WriteLine("NO");    
        }
        else
        {
            stack1 = new Stack<char>();
            bool yesOrNo = true;
            foreach(char x in expression)
            {
                if(x == '[') stack1.Push(x);
                else if(x == '{') stack1.Push(x);
                else if (x == '(') stack1.Push(x);
                else
                {
                    if(stack1.Count > 0)
                    {
                      if((stack1.Peek() != '(' & x == ')') || (stack1.Peek() != '{' & x == '}') || (stack1.Peek() != '[' & x == ']'))
                      {
                        yesOrNo = false;
                      }
                        stack1.Pop();                    
                    }
                    else
                        yesOrNo = false;
                }
            }

            if(yesOrNo && stack1.Count == 0)
                Console.WriteLine("YES");    
            else
                Console.WriteLine("NO");    
        }
    }
  }
}
